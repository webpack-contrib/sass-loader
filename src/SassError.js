class SassError extends Error {
  constructor(sassError) {
    super();

    this.name = "SassError";
    // TODO remove me in the next major release
    this.originalSassError = sassError;

    if (
      typeof sassError.line !== "undefined" ||
      typeof sassError.column !== "undefined"
    ) {
      this.loc = {
        line: sassError.line,
        column: sassError.column,
      };
    }

    // Keep original error if `sassError.formatted` is unavailable
    this.message = `${this.name}: ${
      typeof this.originalSassError.message !== "undefined"
        ? this.originalSassError.message
        : this.originalSassError
    }`;

    if (this.originalSassError.formatted) {
      this.message = `${this.name}: ${this.originalSassError.formatted.replace(
        /^Error: /,
        ""
      )}`;

      // Instruct webpack to hide the JS stack from the console.
      // Usually you're only interested in the SASS stack in this case.
      this.hideStack = true;

      Error.captureStackTrace(this, this.constructor);
    }
  }
}

export default SassError;
