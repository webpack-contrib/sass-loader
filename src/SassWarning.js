class SassWarning extends Error {
  constructor(warning, options) {
    super(warning);

    this.name = "SassWarning";
    this.hideStack = true;

    if (options.span) {
      this.loc = {
        line: options.span.start.line,
        column: options.span.start.column,
      };
    }
  }
}

export default SassWarning;
