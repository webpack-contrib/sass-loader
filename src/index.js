import url from "url";
import path from "path";

import schema from "./options.json";
import {
  getSassImplementation,
  getSassOptions,
  getWebpackImporter,
  getModernWebpackImporter,
  getCompileFn,
  normalizeSourceMap,
  errorFactory,
} from "./utils";

/**
 * The sass-loader makes node-sass and dart-sass available to webpack modules.
 *
 * @this {object}
 * @param {string} content
 */
async function loader(content) {
  const options = this.getOptions(schema);
  const callback = this.async();

  let implementation;

  try {
    implementation = getSassImplementation(this, options.implementation);
  } catch (error) {
    callback(error);

    return;
  }

  const useSourceMap =
    typeof options.sourceMap === "boolean" ? options.sourceMap : this.sourceMap;
  const sassOptions = await getSassOptions(
    this,
    options,
    content,
    implementation,
    useSourceMap
  );
  const shouldUseWebpackImporter =
    typeof options.webpackImporter === "boolean"
      ? options.webpackImporter
      : true;

  if (shouldUseWebpackImporter) {
    const isModernAPI = options.api === "modern";

    if (!isModernAPI) {
      const { includePaths } = sassOptions;

      sassOptions.importer.push(
        getWebpackImporter(this, implementation, includePaths)
      );
    } else {
      sassOptions.importers.push(
        getModernWebpackImporter(this, implementation)
      );
    }
  }

  let compile;

  try {
    compile = getCompileFn(implementation, options);
  } catch (error) {
    callback(error);
    return;
  }

  let result;

  try {
    result = await compile(sassOptions, options);
  } catch (error) {
    // There are situations when the `file`/`span.url` property do not exist
    // Modern API
    if (error.span && typeof error.span.url !== "undefined") {
      this.addDependency(url.fileURLToPath(error.span.url));
    }
    // Legacy API
    else if (typeof error.file !== "undefined") {
      // `node-sass` returns POSIX paths
      this.addDependency(path.normalize(error.file));
    }

    callback(errorFactory(error));

    return;
  }

  let map =
    // Modern API, then legacy API
    result.sourceMap
      ? result.sourceMap
      : result.map
      ? JSON.parse(result.map)
      : null;

  // Modify source paths only for webpack, otherwise we do nothing
  if (map && useSourceMap) {
    map = normalizeSourceMap(map, this.rootContext);
  }

  // Modern API
  if (typeof result.loadedUrls !== "undefined") {
    result.loadedUrls
      .filter((url) => url.protocol === "file:")
      .forEach((includedFile) => {
        const normalizedIncludedFile = url.fileURLToPath(includedFile);

        // Custom `importer` can return only `contents` so includedFile will be relative
        if (path.isAbsolute(normalizedIncludedFile)) {
          this.addDependency(normalizedIncludedFile);
        }
      });
  }
  // Legacy API
  else if (
    typeof result.stats !== "undefined" &&
    typeof result.stats.includedFiles !== "undefined"
  ) {
    result.stats.includedFiles.forEach((includedFile) => {
      const normalizedIncludedFile = path.normalize(includedFile);

      // Custom `importer` can return only `contents` so includedFile will be relative
      if (path.isAbsolute(normalizedIncludedFile)) {
        this.addDependency(normalizedIncludedFile);
      }
    });
  }

  callback(null, result.css.toString(), map);
}

export default loader;
