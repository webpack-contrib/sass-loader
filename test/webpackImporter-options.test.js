import { isSupportedFibers } from "../src/utils";

import {
  compile,
  getCodeFromBundle,
  getCodeFromSass,
  getCompiler,
  getErrors,
  getImplementationsAndAPI,
  getTestId,
  getWarnings,
} from "./helpers";

let Fiber;
const implementations = getImplementationsAndAPI();
const syntaxStyles = ["scss", "sass"];

describe("webpackImporter option", () => {
  beforeAll(async () => {
    if (isSupportedFibers()) {
      const { default: fibers } = await import("fibers");
      Fiber = fibers;
    }
  });

  beforeEach(() => {
    if (isSupportedFibers()) {
      // The `sass` (`Dart Sass`) package modify the `Function` prototype, but the `jest` lose a prototype
      Object.setPrototypeOf(Fiber, Function.prototype);
    }
  });

  implementations.forEach((item) => {
    syntaxStyles.forEach((syntax) => {
      const { name: implementationName, api, implementation } = item;

      // TODO fix me https://github.com/webpack-contrib/sass-loader/issues/774
      if (api === "modern") {
        return;
      }

      it(`not specify ('${implementationName}', '${api}' API, '${syntax}' syntax)`, async () => {
        const testId = getTestId("language", syntax);
        const options = {
          implementation,
          api,
        };
        const compiler = getCompiler(testId, { loader: { options } });
        const stats = await compile(compiler);
        const codeFromBundle = getCodeFromBundle(stats, compiler);
        const codeFromSass = await getCodeFromSass(testId, options);

        expect(codeFromBundle.css).toBe(codeFromSass.css);
        expect(codeFromBundle.css).toMatchSnapshot("css");
        expect(getWarnings(stats)).toMatchSnapshot("warnings");
        expect(getErrors(stats)).toMatchSnapshot("errors");
      });

      it(`false ('${implementationName}', '${api}' API, '${syntax}' syntax)`, async () => {
        const testId = getTestId("language", syntax);
        const options = {
          implementation,
          api,
          webpackImporter: false,
        };
        const compiler = getCompiler(testId, { loader: { options } });
        const stats = await compile(compiler);
        const codeFromBundle = getCodeFromBundle(stats, compiler);
        const codeFromSass = await getCodeFromSass(testId, options);

        expect(codeFromBundle.css).toBe(codeFromSass.css);
        expect(codeFromBundle.css).toMatchSnapshot("css");
        expect(getWarnings(stats)).toMatchSnapshot("warnings");
        expect(getErrors(stats)).toMatchSnapshot("errors");
      });

      it(`true ('${implementationName}', '${api}' API, '${syntax}' syntax)`, async () => {
        const testId = getTestId("language", syntax);
        const options = {
          implementation,
          api,
          webpackImporter: true,
        };
        const compiler = getCompiler(testId, { loader: { options } });
        const stats = await compile(compiler);
        const codeFromBundle = getCodeFromBundle(stats, compiler);
        const codeFromSass = await getCodeFromSass(testId, options);

        expect(codeFromBundle.css).toBe(codeFromSass.css);
        expect(codeFromBundle.css).toMatchSnapshot("css");
        expect(getWarnings(stats)).toMatchSnapshot("warnings");
        expect(getErrors(stats)).toMatchSnapshot("errors");
      });
    });
  });
});
