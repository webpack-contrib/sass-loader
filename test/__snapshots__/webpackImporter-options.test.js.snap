// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webpackImporter option false (dart-sass) (sass): css 1`] = `
"@charset \\"UTF-8\\";
@import \\"./file.css\\";
body {
  font: 100% Helvetica, sans-serif;
  color: #333;
}

nav ul {
  margin: 0;
  padding: 0;
  list-style: none;
}
nav li {
  display: inline-block;
}
nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none;
}

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px;
}

.message, .warning, .error, .success {
  border: 1px solid #ccc;
  padding: 10px;
  color: #333;
}

.success {
  border-color: green;
}

.error {
  border-color: red;
}

.warning {
  border-color: yellow;
}

.foo:before {
  content: \\"\\\\e0c6\\";
}

.bar:before {
  content: \\"∑\\";
}"
`;

exports[`webpackImporter option false (dart-sass) (sass): errors 1`] = `Array []`;

exports[`webpackImporter option false (dart-sass) (sass): warnings 1`] = `Array []`;

exports[`webpackImporter option false (dart-sass) (scss): css 1`] = `
"@charset \\"UTF-8\\";
@import \\"./file.css\\";
body {
  font: 100% Helvetica, sans-serif;
  color: #333;
}

nav ul {
  margin: 0;
  padding: 0;
  list-style: none;
}
nav li {
  display: inline-block;
}
nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none;
}

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px;
}

.foo:before {
  content: \\"\\\\e0c6\\";
}

.bar:before {
  content: \\"∑\\";
}"
`;

exports[`webpackImporter option false (dart-sass) (scss): errors 1`] = `Array []`;

exports[`webpackImporter option false (dart-sass) (scss): warnings 1`] = `Array []`;

exports[`webpackImporter option false (node-sass) (sass): css 1`] = `
"@charset \\"UTF-8\\";
@import url(./file.css);
body {
  font: 100% Helvetica, sans-serif;
  color: #333; }

nav ul {
  margin: 0;
  padding: 0;
  list-style: none; }

nav li {
  display: inline-block; }

nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none; }

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px; }

.message, .success, .error, .warning {
  border: 1px solid #ccc;
  padding: 10px;
  color: #333; }

.success {
  border-color: green; }

.error {
  border-color: red; }

.warning {
  border-color: yellow; }

.foo:before {
  content: \\"\\"; }

.bar:before {
  content: \\"∑\\"; }
"
`;

exports[`webpackImporter option false (node-sass) (sass): errors 1`] = `Array []`;

exports[`webpackImporter option false (node-sass) (sass): warnings 1`] = `Array []`;

exports[`webpackImporter option false (node-sass) (scss): css 1`] = `
"@charset \\"UTF-8\\";
@import url(./file.css);
body {
  font: 100% Helvetica, sans-serif;
  color: #333; }

nav ul {
  margin: 0;
  padding: 0;
  list-style: none; }

nav li {
  display: inline-block; }

nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none; }

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px; }

.foo:before {
  content: \\"\\"; }

.bar:before {
  content: \\"∑\\"; }
"
`;

exports[`webpackImporter option false (node-sass) (scss): errors 1`] = `Array []`;

exports[`webpackImporter option false (node-sass) (scss): warnings 1`] = `Array []`;

exports[`webpackImporter option not specify (dart-sass) (sass): css 1`] = `
"@charset \\"UTF-8\\";
@import \\"./file.css\\";
body {
  font: 100% Helvetica, sans-serif;
  color: #333;
}

nav ul {
  margin: 0;
  padding: 0;
  list-style: none;
}
nav li {
  display: inline-block;
}
nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none;
}

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px;
}

.message, .warning, .error, .success {
  border: 1px solid #ccc;
  padding: 10px;
  color: #333;
}

.success {
  border-color: green;
}

.error {
  border-color: red;
}

.warning {
  border-color: yellow;
}

.foo:before {
  content: \\"\\\\e0c6\\";
}

.bar:before {
  content: \\"∑\\";
}"
`;

exports[`webpackImporter option not specify (dart-sass) (sass): errors 1`] = `Array []`;

exports[`webpackImporter option not specify (dart-sass) (sass): warnings 1`] = `Array []`;

exports[`webpackImporter option not specify (dart-sass) (scss): css 1`] = `
"@charset \\"UTF-8\\";
@import \\"./file.css\\";
body {
  font: 100% Helvetica, sans-serif;
  color: #333;
}

nav ul {
  margin: 0;
  padding: 0;
  list-style: none;
}
nav li {
  display: inline-block;
}
nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none;
}

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px;
}

.foo:before {
  content: \\"\\\\e0c6\\";
}

.bar:before {
  content: \\"∑\\";
}"
`;

exports[`webpackImporter option not specify (dart-sass) (scss): errors 1`] = `Array []`;

exports[`webpackImporter option not specify (dart-sass) (scss): warnings 1`] = `Array []`;

exports[`webpackImporter option not specify (node-sass) (sass): css 1`] = `
"@charset \\"UTF-8\\";
@import url(./file.css);
body {
  font: 100% Helvetica, sans-serif;
  color: #333; }

nav ul {
  margin: 0;
  padding: 0;
  list-style: none; }

nav li {
  display: inline-block; }

nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none; }

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px; }

.message, .success, .error, .warning {
  border: 1px solid #ccc;
  padding: 10px;
  color: #333; }

.success {
  border-color: green; }

.error {
  border-color: red; }

.warning {
  border-color: yellow; }

.foo:before {
  content: \\"\\"; }

.bar:before {
  content: \\"∑\\"; }
"
`;

exports[`webpackImporter option not specify (node-sass) (sass): errors 1`] = `Array []`;

exports[`webpackImporter option not specify (node-sass) (sass): warnings 1`] = `Array []`;

exports[`webpackImporter option not specify (node-sass) (scss): css 1`] = `
"@charset \\"UTF-8\\";
@import url(./file.css);
body {
  font: 100% Helvetica, sans-serif;
  color: #333; }

nav ul {
  margin: 0;
  padding: 0;
  list-style: none; }

nav li {
  display: inline-block; }

nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none; }

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px; }

.foo:before {
  content: \\"\\"; }

.bar:before {
  content: \\"∑\\"; }
"
`;

exports[`webpackImporter option not specify (node-sass) (scss): errors 1`] = `Array []`;

exports[`webpackImporter option not specify (node-sass) (scss): warnings 1`] = `Array []`;

exports[`webpackImporter option true (dart-sass) (sass): css 1`] = `
"@charset \\"UTF-8\\";
@import \\"./file.css\\";
body {
  font: 100% Helvetica, sans-serif;
  color: #333;
}

nav ul {
  margin: 0;
  padding: 0;
  list-style: none;
}
nav li {
  display: inline-block;
}
nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none;
}

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px;
}

.message, .warning, .error, .success {
  border: 1px solid #ccc;
  padding: 10px;
  color: #333;
}

.success {
  border-color: green;
}

.error {
  border-color: red;
}

.warning {
  border-color: yellow;
}

.foo:before {
  content: \\"\\\\e0c6\\";
}

.bar:before {
  content: \\"∑\\";
}"
`;

exports[`webpackImporter option true (dart-sass) (sass): errors 1`] = `Array []`;

exports[`webpackImporter option true (dart-sass) (sass): warnings 1`] = `Array []`;

exports[`webpackImporter option true (dart-sass) (scss): css 1`] = `
"@charset \\"UTF-8\\";
@import \\"./file.css\\";
body {
  font: 100% Helvetica, sans-serif;
  color: #333;
}

nav ul {
  margin: 0;
  padding: 0;
  list-style: none;
}
nav li {
  display: inline-block;
}
nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none;
}

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px;
}

.foo:before {
  content: \\"\\\\e0c6\\";
}

.bar:before {
  content: \\"∑\\";
}"
`;

exports[`webpackImporter option true (dart-sass) (scss): errors 1`] = `Array []`;

exports[`webpackImporter option true (dart-sass) (scss): warnings 1`] = `Array []`;

exports[`webpackImporter option true (node-sass) (sass): css 1`] = `
"@charset \\"UTF-8\\";
@import url(./file.css);
body {
  font: 100% Helvetica, sans-serif;
  color: #333; }

nav ul {
  margin: 0;
  padding: 0;
  list-style: none; }

nav li {
  display: inline-block; }

nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none; }

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px; }

.message, .success, .error, .warning {
  border: 1px solid #ccc;
  padding: 10px;
  color: #333; }

.success {
  border-color: green; }

.error {
  border-color: red; }

.warning {
  border-color: yellow; }

.foo:before {
  content: \\"\\"; }

.bar:before {
  content: \\"∑\\"; }
"
`;

exports[`webpackImporter option true (node-sass) (sass): errors 1`] = `Array []`;

exports[`webpackImporter option true (node-sass) (sass): warnings 1`] = `Array []`;

exports[`webpackImporter option true (node-sass) (scss): css 1`] = `
"@charset \\"UTF-8\\";
@import url(./file.css);
body {
  font: 100% Helvetica, sans-serif;
  color: #333; }

nav ul {
  margin: 0;
  padding: 0;
  list-style: none; }

nav li {
  display: inline-block; }

nav a {
  display: block;
  padding: 6px 12px;
  text-decoration: none; }

.box {
  -webkit-border-radius: 10px;
  -moz-border-radius: 10px;
  -ms-border-radius: 10px;
  border-radius: 10px; }

.foo:before {
  content: \\"\\"; }

.bar:before {
  content: \\"∑\\"; }
"
`;

exports[`webpackImporter option true (node-sass) (scss): errors 1`] = `Array []`;

exports[`webpackImporter option true (node-sass) (scss): warnings 1`] = `Array []`;
